[iOS 17] Software keyboard overlaps position: fixed; text fields near the bottom of the viewport

whsieh opened the pull request:
#### 6669d12b7b43d2d36a549d9fcad0a1577b66a1f4
<pre>
[iOS 17] Software keyboard overlaps position: fixed; text fields near the bottom of the viewport
<a href="https://bugs.webkit.org/show_bug.cgi?id=258828">https://bugs.webkit.org/show_bug.cgi?id=258828</a>
rdar://109127515

Reviewed by Aditya Keerthi.

On iOS 17, in the case where out-of-process keyboard is enabled and the software keyboard is shown,
we end up getting two sets of `&quot;KeyboardWillShow&quot;` -&gt; `&quot;KeyboardDidShow&quot;` notifications when the
keyboard animates up, after reloading input views. When the first set of notifications is dispatched
underneath the  call to `-reloadInputViews`, the keyboard hasn&apos;t yet become full height, and so our
attempts to zoom to reveal the focused element using the current input view bounds will fail.

I filed &lt;rdar://111704216&gt; for UIKit to investigate restoring pre-iOS-17 behavior, with respect to
dispatching these intermediate keyboard notifications; in the meantime, this patch works around this
behavior change by deferring the call to `-_zoomToRevealFocusedElement` until the first
`&quot;KeyboardWillShow&quot;` notification arrives *after* `-reloadInputViews` has already been invoked, in
the case where OOP keyboard is enabled and the keyboard is full height.

We limit this to the case of OOP keyboard because when that feature is disabled, we receive a
&quot;KeyboardWillShow&quot; notification underneath the call to `-reloadInputViews` that already contains the
final keyboard height; furthermore, we limit this to the case where the keyboard is full height,
because when OOP keyboard is enabled and the keyboard is minimized (e.g. the hardware keyboard is
attached), we only get a single set of keyboard appearance notifications.

* LayoutTests/fast/forms/ios/scroll-to-reveal-fixed-input-expected.txt: Added.
* LayoutTests/fast/forms/ios/scroll-to-reveal-fixed-input.html: Added.

Add a new layout test to exercise keyboard scrolling, in the case where:

1. We&apos;re showing the software keyboard.
2. The focused element that would be overlapped by the keyboard is in a fixed position container.
3. The page itself is scrollable.

...and verify that we successfully scroll such that the bottom of the caret rect is above the top of
the keyboard (input view bounds).

* LayoutTests/resources/ui-helper.js:
(window.UIHelper.getUICaretViewRectInGlobalCoordinates.return.new.Promise.):
(window.UIHelper.getUICaretViewRectInGlobalCoordinates.return.new.Promise):
(window.UIHelper.getUICaretViewRectInGlobalCoordinates):

Add support for a new `UIHelper` method to grab the selection caret rect in web view coordinates.
The &quot;global&quot; here is consistent terminology used elsewhere in `UIScriptController`.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _elementDidFocus:userIsInteracting:blurPreviousNode:activityStateChanges:userObject:]):

See comments above for more details.

* Tools/TestRunnerShared/UIScriptContext/Bindings/UIScriptController.idl:
* Tools/TestRunnerShared/UIScriptContext/UIScriptController.h:
(WTR::UIScriptController::selectionCaretViewRectInGlobalCoordinates const):
* Tools/WebKitTestRunner/ios/UIScriptControllerIOS.h:
* Tools/WebKitTestRunner/ios/UIScriptControllerIOS.mm:
(WTR::UIScriptControllerIOS::selectionCaretViewRect const):
(WTR::UIScriptControllerIOS::selectionCaretViewRectInGlobalCoordinates const):

Canonical link: <a href="https://commits.webkit.org/265741@main">https://commits.webkit.org/265741@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0d371d12874313f39a67484af6192604e172f2b7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/11764 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/11963 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/12331 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/13408 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/11204 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/11783 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/14352 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/11944 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14068 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/11928 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/12761 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/9985 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/13829 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10053 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/10679 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/17809 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11130 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/10834 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14004 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/11240 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/9280 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10415 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/10558 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2832 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/14698 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/11097 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas reviewed (approved):
LGTM

aprotyas commented on Tools/WebKitTestRunner/ios/UIScriptControllerIOS.h:
```diff
@@ -187,6 +190,8 @@ class UIScriptControllerIOS final : public UIScriptControllerCocoa {
     UITextSelectionDisplayInteraction *textSelectionDisplayInteraction() const;
 #endif
+
+    JSObjectRef selectionCaretViewRect(id<UICoordinateSpace>) const;
```
Optional nit: place this around the other `selectionCaretViewRect*` functions?

whsieh replied:
Sure! (I placed it near the bottom because it's only an internal helper method rather than an overridden method on UISC, but either way works for me)

aprotyas replied:
Yeah that makes sense, I was going to suggest `private`-ing it, then I see everything in this class is `private` anyway!

whsieh committed:
[iOS 17] Software keyboard overlaps position: fixed; text fields near the bottom of the viewport
https://bugs.webkit.org/show_bug.cgi?id=258828
rdar://109127515

Reviewed by Aditya Keerthi.

On iOS 17, in the case where out-of-process keyboard is enabled and the software keyboard is shown,
we end up getting two sets of `"KeyboardWillShow"` -> `"KeyboardDidShow"` notifications when the
keyboard animates up, after reloading input views. When the first set of notifications is dispatched
underneath the  call to `-reloadInputViews`, the keyboard hasn't yet become full height, and so our
attempts to zoom to reveal the focused element using the current input view bounds will fail.

I filed <rdar://111704216> for UIKit to investigate restoring pre-iOS-17 behavior, with respect to
dispatching these intermediate keyboard notifications; in the meantime, this patch works around this
behavior change by deferring the call to `-_zoomToRevealFocusedElement` until the first
`"KeyboardWillShow"` notification arrives *after* `-reloadInputViews` has already been invoked, in
the case where OOP keyboard is enabled and the keyboard is full height.

We limit this to the case of OOP keyboard because when that feature is disabled, we receive a
"KeyboardWillShow" notification underneath the call to `-reloadInputViews` that already contains the
final keyboard height; furthermore, we limit this to the case where the keyboard is full height,
because when OOP keyboard is enabled and the keyboard is minimized (e.g. the hardware keyboard is
attached), we only get a single set of keyboard appearance notifications.

* LayoutTests/fast/forms/ios/scroll-to-reveal-fixed-input-expected.txt: Added.
* LayoutTests/fast/forms/ios/scroll-to-reveal-fixed-input.html: Added.

Add a new layout test to exercise keyboard scrolling, in the case where:

1. We're showing the software keyboard.
2. The focused element that would be overlapped by the keyboard is in a fixed position container.
3. The page itself is scrollable.

...and verify that we successfully scroll such that the bottom of the caret rect is above the top of
the keyboard (input view bounds).

* LayoutTests/resources/ui-helper.js:
(window.UIHelper.getUICaretViewRectInGlobalCoordinates.return.new.Promise.):
(window.UIHelper.getUICaretViewRectInGlobalCoordinates.return.new.Promise):
(window.UIHelper.getUICaretViewRectInGlobalCoordinates):

Add support for a new `UIHelper` method to grab the selection caret rect in web view coordinates.
The "global" here is consistent terminology used elsewhere in `UIScriptController`.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _elementDidFocus:userIsInteracting:blurPreviousNode:activityStateChanges:userObject:]):

See comments above for more details.

* Tools/TestRunnerShared/UIScriptContext/Bindings/UIScriptController.idl:
* Tools/TestRunnerShared/UIScriptContext/UIScriptController.h:
(WTR::UIScriptController::selectionCaretViewRectInGlobalCoordinates const):
* Tools/WebKitTestRunner/ios/UIScriptControllerIOS.h:
* Tools/WebKitTestRunner/ios/UIScriptControllerIOS.mm:
(WTR::UIScriptControllerIOS::selectionCaretViewRect const):
(WTR::UIScriptControllerIOS::selectionCaretViewRectInGlobalCoordinates const):

Canonical link: https://commits.webkit.org/265741@main
